### ISSUE TITLE
z-image training and resuming

### ISSUE THREAD

**Issue by GraftingRayman** [kohya-ss/musubi-tuner#776]
Anyone had any luck resuming a broken training session?

I am using --save_state as a parameter, when using --resume path to the epoch savestate folder, it resumes from step 0 of the total original steps

i am using the following command:

accelerate launch --num_cpu_threads_per_process 1 --mixed_precision bf16 -m musubi_tuner.zimage_train_network --dit=E:\models\z_image_de_turbo_v1_bf16.safetensors --vae=E:\models\zimage_vae.safetensors --text_encoder=E:\models\zImage_textEncoder.safetensors --dataset_config=I:\Source Images\zimage-nottrained\E_01\10_E_01\config.toml --sdpa --mixed_precision=bf16 --timestep_sampling=shift --weighting_scheme=none --discrete_flow_shift=2.0 --optimizer_type=adamw8bit --learning_rate=1e-4 --gradient_checkpointing --gradient_checkpointing_cpu_offload --blocks_to_swap=20 --fp8_base --fp8_scaled --fp8_llm --max_data_loader_n_workers=2 --persistent_data_loader_workers --network_module=networks.lora_zimage --network_dim=32 --max_train_epochs=16 --save_every_n_epochs=1 --sample_every_n_steps=760 --seed=42 --output_dir=I:\Source Images\zimage-nottrained\E_01 --output_name=zimage_10_E_01 --resume=I:\Source Images\zimage-nottrained\E_01\zimage_10_E_01-000001-state --save_state --sample_prompts=e:\models\sample_prompts.txt

76 Images, 10 Repeats and 16 Epochs

Total steps is 12,160, resuming after Epoch 1 should leave 11,400 but its still doing 12,160

The same logic for Flux works fine

**Comment by GraftingRayman** [comment]
Went back and compared Flux save state vs Z-Image save state, there is a file called train_state.json which contains:

epoch 1:
{"current_epoch": 1, "current_step": 300} 

epoch 2:
{"current_epoch": 2, "current_step": 600} 

This file is not there for z-image, not even sure if its required for z-image, but could this be why resumption fails for z-image?

I placed a custom train_state.json and the training script I believe ignores it.

**Comment by GraftingRayman** [comment]
Ok, its working now, added the missing code to hv_train_network.py (deepseek)

Now it resumes at the state selected to resume from

If need be, you can check my fork for the differences

        # ✅ ADD THESE: Shared memory variables for tracking epoch and step
        current_step = Value('i', 0)  # Shared counter for steps
        current_epoch = Value('i', 0)  # Shared counter for epochs
        steps_from_state = None  # Will be set by load_model_hook
        epochs_from_state = None  # Will be set by load_model_hook



        # before resuming make hook for saving/loading to save/load the network weights only
        def save_model_hook(models, weights, output_dir):
            # pop weights of other models than network to save only network weights
            # only main process or deepspeed https://github.com/huggingface/diffusers/issues/2606
            if accelerator.is_main_process:  # or args.deepspeed:
                remove_indices = []
                for i, model in enumerate(models):
                    if not isinstance(model, type(accelerator.unwrap_model(network))):
                        remove_indices.append(i)
                for i in reversed(remove_indices):
                    if len(weights) > i:
                        weights.pop(i)
                # print(f"save model hook: {len(weights)} weights will be saved")
                
                # ✅ ADD THIS: Save current epoch and step to train_state.json
                train_state_file = os.path.join(output_dir, "train_state.json")
                # +1 is needed because the state is saved before current_step is set from global_step
                logger.info(f"save train state to {train_state_file} at epoch {current_epoch.value} step {current_step.value+1}")
                with open(train_state_file, "w", encoding="utf-8") as f:
                    json.dump({"current_epoch": current_epoch.value, "current_step": current_step.value + 1}, f)
    
        def load_model_hook(models, input_dir):
            # remove models except network
            remove_indices = []
            for i, model in enumerate(models):
                if not isinstance(model, type(accelerator.unwrap_model(network))):
                    remove_indices.append(i)
            for i in reversed(remove_indices):
                models.pop(i)
            # print(f"load model hook: {len(models)} models will be loaded")
    
            # ✅ ADD THIS: Load current epoch and step from train_state.json
            nonlocal steps_from_state, epochs_from_state
            train_state_file = os.path.join(input_dir, "train_state.json")
            if os.path.exists(train_state_file):
                with open(train_state_file, "r", encoding="utf-8") as f:
                    data = json.load(f)
                steps_from_state = data["current_step"]
                epochs_from_state = data["current_epoch"]
                logger.info(f"load train state from {train_state_file}: {data}")


        # ✅ USE LOADED STATE TO ADJUST TRAINING START POSITIONS
        if steps_from_state is not None:
            global_step = steps_from_state
            # Update accelerator's internal step counter
            accelerator.step = global_step
            logger.info(f"Resuming from step: {global_step}")
        else:
            global_step = 0
    
        if epochs_from_state is not None:
            epoch_to_start = epochs_from_state
            logger.info(f"Resuming from epoch: {epoch_to_start}")
        else:
            epoch_to_start = 0

**Comment by GraftingRayman** [comment]
So found another issue, when resumption takes place, the training goes ahead as it should do.

The issue I have found, the sample images look like its using the model from the save state, ie epoch 3 state was selected, for all the sample images generated it was using the model for that epoch. Samples generated were identical.

The training itself was successful, each epoch when used as a lora in my workflow showed better results in generating the likeness from the previous.

**Comment by taxuew** [comment]
Hello, I am running a fine-tuning job using eight H20 GPUs and I am experiencing an Out-of-Memory (OOM) error. However, I can successfully fine-tune the same model on a single H20 card without this issue. Have you encountered this problem before, and how did you resolve it?"
